infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationManager.java

```diff
@@ -0,0 +1,300 @@
+            log.error("Unrecoverable exception when attempting to connect to remote cluster.", e);
+            throw e;
+        }
+    }
+
+    /**
+     * Stop Log Replication to a specific standby Cluster
+     */
+    private void stopLogReplicationRuntime(String remoteClusterId) {
+        CorfuLogReplicationRuntime logReplicationRuntime = runtimeToRemoteCluster.get(remoteClusterId);
+        if (logReplicationRuntime != null) {
+            log.info("Stop log replication runtime to remote cluster id={}", remoteClusterId);
+            logReplicationRuntime.stop();
+            runtimeToRemoteCluster.remove(remoteClusterId);
+        } else {
+            log.warn("Runtime not found to remote cluster {}", remoteClusterId);
+        }
+    }
+
+    /**
+     * The notification of change of adding/removing standby's without epoch change.
+     *
+     * @param newConfig has the same topologyConfigId as the current config
+     */
+    public void processStandbyChange(TopologyDescriptor newConfig) {
+        if (newConfig.getTopologyConfigId() != topologyDescriptor.getTopologyConfigId()) {
+            log.error("Detected changes in the topology. The new topology descriptor {} doesn't have the same " +
+                    "siteConfigId as the current one {}", newConfig, topologyDescriptor);
+            return;
+        }
+
+        Map<String, ClusterDescriptor> newStandbys = newConfig.getStandbyClusters();
+        Map<String, ClusterDescriptor> currentStandbys = topologyDescriptor.getStandbyClusters();
+        newStandbys.keySet().retainAll(currentStandbys.keySet());
+        Set<String> standbysToRemove = currentStandbys.keySet();
+        standbysToRemove.removeAll(newStandbys.keySet());
+
+        /*
+         * Remove standbys that are not in the new config
+         */
+        for (String siteID : standbysToRemove) {
+            stopLogReplicationRuntime(siteID);
+            topologyDescriptor.removeStandbySite(siteID);
+        }
+
+        //Start the standbys that are in the new config but not in the current config
+        for (String clusterId : newConfig.getStandbyClusters().keySet()) {
+            if (runtimeToRemoteCluster.get(clusterId) == null) {
+                ClusterDescriptor clusterInfo = newConfig.getStandbyClusters().get(clusterId);
+                topologyDescriptor.addStandbySite(clusterInfo);
+                startLogReplicationRuntime(clusterInfo);
+            }
+        }
+    }
+
+    /**
+     * Query max stream tail for all streams to be replicated.
+     *
+     * @return max tail of all relevant streams.
+     */
+    private long queryStreamTail() {
+        Set<String> streamsToReplicate = logReplicationConfig.getStreamsToReplicate();
+        long maxTail = Address.NON_ADDRESS;
+        for (String s : streamsToReplicate) {
+            UUID currentUUID = CorfuRuntime.getStreamID(s);
+            Map<UUID, Long> tailMap = corfuRuntime.getAddressSpaceView().getAllTails().getStreamTails();
```

Comment:
Query tailMap multiple times is not efficient.

Reply:
@xiaoqin2012 please address this...

Reply:
We can move the tailMap query above the for loop. Good catch!